Qiskit/qiskit-terra
# Issue 3750 wrap up

### Summary
Closes #3750 

### Details and comments


Minor API changes
- `operands` returns a tuple rather than a list (better type hints)
- add `id` for insts and pulses and make default `name` `None`

Bug fix
- fix `__repr__` bug for `Discriminator` and `Kernel` config

Docs
- Overall improved pulse documentation

Misc / implementation changes
- clean up magic methods and their inheritance for `Pulse`+subclasses and `Instruction`+subclasses
- `PulseCommand` class wasn't used anywhere so I removed the file :D

Tests
- `test_instructions.py`: New instruction tests and migrate Acquire tests
- new file `test_pulse_lib.py`: Move tests for `ParametricPulses` and `SamplePulse` to this since they are no longer `Command`s
- update tests to migrate to new pattern or use `self.assertWarns(DeprecationWarning)` to catch test warnings
- add a few tests throughout (e.g. test_assembler)
